lib/Dialect/SPIRV/Transforms/DecorateSPIRVCompositeTypeLayoutPass.cpp

```diff
@@ -0,0 +1,318 @@
+/// alignment of the array or matrix as defined above.
+/// 3. The ArrayStride, MatrixStride, and Offset decorations must be large
+/// enough to hold the size of the objects they affect (that is, specifying
+/// overlap is invalid).
+class VulkanLayoutUtils {
+public:
+  using Alignment = uint64_t;
+
+  /// Returns a new type with layout info. Assigns the type size in bytes to the
+  /// `size`. Assigns the type alignment in bytes to the `alignment`.
+  static Type decorateType(spirv::StructType structType,
+                           spirv::StructType::LayoutInfo &size,
+                           Alignment &alignment);
+  /// Checks whether a type is legal in terms of Vulkan layout info
+  /// decoration. A type is dynamically legal if it's a composite type in the
+  /// StorageBuffer, PhysicalStorageBuffer, Uniform, and PushConstant Storage
+  /// Classes without layout informtation.
+  static bool isLegalType(Type type);
+
+private:
+  static Type decorateType(Type type, spirv::StructType::LayoutInfo &size,
+                           Alignment &alignment);
+  static Type decorateType(VectorType compositeType,
+                           spirv::StructType::LayoutInfo &size,
+                           Alignment &aligment);
+  static Type decorateType(spirv::ArrayType compositeType,
+                           spirv::StructType::LayoutInfo &size,
+                           Alignment &aligment);
+  /// Calculates the alignment for the given scalar type.
+  static Alignment getScalarTypeAlignment(Type scalarType);
+};
+
+Type VulkanLayoutUtils::decorateType(
+    spirv::StructType structType, spirv::StructType::LayoutInfo &size,
+    VulkanLayoutUtils::Alignment &alignment) {
+  if (!structType.getNumElements()) {
+    return structType;
+  }
+
+  llvm::SmallVector<Type, 4> memberTypes;
+  llvm::SmallVector<spirv::StructType::LayoutInfo, 4> layoutInfo;
+  llvm::SmallVector<spirv::StructType::MemberDecorationInfo, 4>
+      memberDecorations;
+
+  spirv::StructType::LayoutInfo structMemberOffset = 0;
+  spirv::StructType::LayoutInfo memberSize = 0;
+  VulkanLayoutUtils::Alignment maxMemberAlignment = 1;
+  VulkanLayoutUtils::Alignment memberAlignment = 1;
+
+  for (uint32_t i = 0, e = structType.getNumElements(); i < e; ++i) {
+    auto memberType = VulkanLayoutUtils::decorateType(
+        structType.getElementType(i), memberSize, memberAlignment);
+    structMemberOffset = llvm::alignTo(structMemberOffset, memberAlignment);
+    memberTypes.push_back(memberType);
+    layoutInfo.push_back(structMemberOffset);
+    // According to the Vulkan spec:
+    // "A structure has a base alignment equal to the largest base alignment of
+    // any of its members."
+    structMemberOffset += memberSize;
+    maxMemberAlignment = std::max(maxMemberAlignment, memberAlignment);
+    memberAlignment = 1;
+    memberSize = 0;
+  }
+
+  size = llvm::alignTo(structMemberOffset, maxMemberAlignment);
```

Comment:
Is this correct? Why we are adjusting the size with alignment? (We need that when placing the struct inside another struct, but that's adjusting the offset instead of the struct size itself and it should be handled by the outer struct's member layout logic, right?)

Reply:
@antiagainst thanks for review, I've looked at this rule https://github.com/microsoft/DirectXShaderCompiler/blob/3e302950dd507e9c433705111ac0de1b143d6d1b/tools/clang/lib/SPIRV/AlignmentSizeCalculator.cpp#L324

but I might be wrong with an interpretation, please fix me in this case.

If we do not align struct size to the max alignment of its member, we will get:
` !spv.ptr<!spv.struct<!spv.struct<!spv.struct<i1 [0], i64 [8]> [0], i1 [16]> [0], i1 [17]>, Uniform>`

with alignment for struct size:
` !spv.ptr<!spv.struct<!spv.struct<!spv.struct<i1 [0], i64 [8]> [0], i1 [16]> [0], i1 [24]>, Uniform>`

It's similiar to c++ rules I think, but I'm not sure that rules are applicable for spir-v alignment.
```
struct A {
  bool a;
  uint64_t b;
};

struct B {
  A a;
  bool b;
};

struct C {
  B b;
  bool c;
};

sizeof (A) == 16
sizeof (B) == 24
sizoef (C) == 32
```

Thanks.

Reply:
Agree with @denis0x0D here. The only way to make sure that you satisfy the alignment properties of all elements of a struct, is to make sure that the struct itself is aligned to the maximum alignment of any element of the struct.

Reply:
Ah yes, sorry forgot about that rule. :) Could you quote the spec here? "The Offset decoration of a member must not place it between the end of a structure or an array and the next multiple of the alignment of that structure or array."
 
@MaheshRavishankar: it's a slightly different issue here. The struct member alignment is satisfied at Line 98. Here we are talking about the element following a struct inside another struct.